Bottom: 43a101856751ff43466213e6cd56a76c84615cf3
Top:    c7d7650890337a7061e170c91422729b35a21916
Author: Jason Wang <jasowang@redhat.com>
Date:   2018-07-18 20:25:30 +0800

Refresh of virtio-introduce-packed-ring

---

diff --git a/include/uapi/linux/virtio_config.h b/include/uapi/linux/virtio_config.h
index 1196e1c..b2063b6 100644
--- a/include/uapi/linux/virtio_config.h
+++ b/include/uapi/linux/virtio_config.h
@@ -75,7 +75,11 @@
  */
 #define VIRTIO_F_IOMMU_PLATFORM		33
 
+<<<<<<< current
 /* This feature indicates support for the packed virtqueue layout. */
+=======
+/* Packed virtqueue layout */
+>>>>>>> patched
 #define VIRTIO_F_RING_PACKED		34
 
 /*
diff --git a/include/uapi/linux/virtio_ring.h b/include/uapi/linux/virtio_ring.h
index 0254a2b..bba700a 100644
--- a/include/uapi/linux/virtio_ring.h
+++ b/include/uapi/linux/virtio_ring.h
@@ -43,6 +43,8 @@
 #define VRING_DESC_F_WRITE	2
 /* This means the buffer contains a list of buffer descriptors. */
 #define VRING_DESC_F_INDIRECT	4
+#define VRING_DESC_F_AVAIL	(1 << 7)
+#define VRING_DESC_F_USED	(1 << 15)
 
 /* Mark a descriptor as available or used. */
 #define VRING_DESC_F_AVAIL	(1ul << 7)
@@ -77,6 +79,36 @@
  * at the end of the used ring. Guest should ignore the used->flags field. */
 #define VIRTIO_RING_F_EVENT_IDX		29
 
+struct vring_packed_desc {
+	/* Buffer Address. */
+	__virtio64 addr;
+	/* Buffer Length. */
+	__virtio32 len;
+	/* Buffer ID. */
+	__virtio16 id;
+	/* The flags depending on descriptor type. */
+	__virtio16 flags;
+};
+
+/* Enable events */
+#define VRING_EVENT_F_ENABLE 0x0
+/* Disable events */
+#define VRING_EVENT_F_DISABLE 0x1
+/*
+ * Enable events for a specific descriptor
+ * (as specified by Descriptor Ring Change Event Offset/Wrap Counter).
+ * Only valid if VIRTIO_F_RING_EVENT_IDX has been negotiated.
+ */
+#define VRING_EVENT_F_DESC 0x2
+/* The value 0x3 is reserved */
+
+struct vring_packed_desc_event {
+	/* Descriptor Ring Change Event Offset and Wrap Counter */
+	__virtio16 off_wrap;
+	/* Descriptor Ring Change Event Flags */
+	__virtio16 flags;
+};
+
 /* Virtio ring descriptors: 16 bytes.  These can chain together via "next". */
 struct vring_desc {
 	/* Address (guest-physical). */
